src/db/sysdb_search.c

```diff
@@ -545,8 +553,19 @@ static char *enum_filter(TALLOC_CTX *mem_ctx,
         filter = talloc_asprintf(tmp_ctx, "(&%s", base_filter);
 
         if (filter != NULL && attr != NULL && attr_filter != NULL) {
-            filter = talloc_asprintf_append(filter, "(%s=%s)",
-                                            attr, attr_filter);
+            if (domain == NULL || !is_sysdb_name(attr)) {
+                value = attr_filter;
+            } else {
+                value = sss_create_internal_fqname(tmp_ctx, attr_filter, domain);
+                if (value == NULL) {
+                    filter = NULL;
+                }
+            }
+
+            if (filter != NULL) {
+                filter = talloc_asprintf_append(filter, "(%s=%s)",
+                                                attr, value);
+            }
```

Comment:
This part of the code is difficult to understand. First you set the value of `filter`, next you check that it isn't null alongside other variables, then if a condition regarding another variable is met you nullify it. Finally, if `filter` isn't null you assign it another value.

There may be no other way to implement it, but I fear that we will introduce some bug in such a convoluted code.

Reply:
I simplified this function by creating a subfunction.

Reply:
It's clearer now. Thanks!